src/core/qgstiledownloadmanager.h

```diff
@@ -0,0 +1,292 @@
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#ifndef QGSTILEDOWNLOADMANAGER_H
+#define QGSTILEDOWNLOADMANAGER_H
+
+#define SIP_NO_FILE
+
+#include <QTimer>
+#include <QThread>
+#include <QMutex>
+
+#include <QNetworkAccessManager>
+#include <QNetworkReply>
+
+#include "qgis_core.h"
+
+class QgsTileDownloadManager;
+
+/**
+ * \ingroup core
+ * Reply object for tile download manager requests returned from calls to QgsTileDownloadManager::get().
+ *
+ * When the underlying network request has finished (with success or failure), the finished() signal
+ * gets emitted.
+ *
+ * It is OK to delete this object before the request has finished - the request will not be aborted,
+ * the download manager will finish the download (as it may be needed soon afterwards).
+ *
+ * \since QGIS 3.18
+ */
+class CORE_EXPORT QgsTileDownloadManagerReply : public QObject
+{
+    Q_OBJECT
+  public:
+    ~QgsTileDownloadManagerReply();
+
+    //! Returns whether the reply has already finished (with success/failure)
+    bool hasFinished() const { return mHasFinished; }
+    //! Returns binary data returned in the reply (only valid when already finished)
+    QByteArray data() const { return mData; }
+    //! Returns error code (only valid when already finished)
+    QNetworkReply::NetworkError error() const { return mError; }
+    //! Returns error string (only valid when already finished)
+    QString errorString() const { return mErrorString; }
+
+    //! Returns the original request for this reply object
+    QNetworkRequest request() const { return mRequest; }
+
+  signals:
+    //! Emitted when the reply has finished (either with a success or with a failure)
+    void finished();
+
+  private slots:
+    void requestFinished( QByteArray data, QNetworkReply::NetworkError error, const QString &errorString );
+
+  private:
+    QgsTileDownloadManagerReply( QgsTileDownloadManager *manager, const QNetworkRequest &request );
+
+    friend class QgsTileDownloadManager;  // allows creation of new instances from the manager
+
+  private:
+    //! "parent" download manager of this reply
+    QgsTileDownloadManager *mManager = nullptr;
+    QNetworkRequest mRequest;
+    bool mHasFinished = false;
+    QByteArray mData;
+    QNetworkReply::NetworkError mError = QNetworkReply::NoError;
+    QString mErrorString;
+};
+
+
+/// @cond PRIVATE
```

Comment:
suggestion: wouldn't it be better to move all these private classes to new `qgstiledownloadmanager_p.h/cpp` files?